### PR TITLE
feat(game-servers): game server provider interface

### PR DESCRIPTION
**Pull request by garrappachc**
This PR abstracts out the gameserver provider by adding the `GameServerProvider` class. The providers can be implemented as separate modules:
* implement the `GameServerProvider`
* add MongoDB discriminator in the provider module `providers`, e.g.
```ts
const staticGameServerModelProvider: Provider = {
  provide: getModelToken(StaticGameServer.name),
  useFactory: (gameServerModel: Model<GameServerDocument>) =>
    gameServerModel.discriminator('static', staticGameServerSchema),
  inject: [getModelToken(GameServer.name)],
};
```
* register the provider to the `GameServersService`:
```ts
async onModuleInit() {
  this.gameServersService.registerProvider(this);
}
```

Currently implemented gameserver provider is represented as `static`, as it gives access to the statically-configured gameservers.

### CODE REVIEW

**Comment by c43721** [comment]
FYI e2e failing on game server service

File: src/game-servers/models/game-server.ts [hunk 1]
```diff
@@ -21,39 +24,27 @@ export class GameServer extends MongooseDocument {
...
   @Prop({ type: Types.ObjectId, ref: 'Game' })
   game?: Types.ObjectId; // currently running game
 
-  @Prop({ default: () => new Date() })
-  lastHeartbeatAt?: Date;
+  /**
+   * Create a new RCON connection to the gameserver.
+   */
+  async rcon(): Promise<Rcon> {
+    throw new Error('not implemented');
```
Comment (c43721) [inline src/game-servers/models/game-server.ts]: Can we mark this method as abstract? We're using inherience anyways, unless nestjs mongoose is dumb.

**Comment by garrappachc** [comment]
> FYI e2e failing on game server service

Yeah, I'm aware ;) This PR needs some polishing still, not quite done with it yet.

File: src/game-servers/providers/static-game-server/controllers/static-game-servers.controller.ts [hunk 2]
```diff
@@ -22,19 +21,22 @@ import { Secret } from '@/auth/decorators/secret.decorator';
 import { GameServerHeartbeat } from '../dto/game-server-heartbeat';
 import { RealIp } from 'nestjs-real-ip';
 import { SecretPurpose } from '@/auth/secret-purpose';
+import { StaticGameServersService } from '../services/static-game-servers.service';
+import { GameServersService } from '@/game-servers/services/game-servers.service';
 
-@Controller('game-servers')
-export class GameServersController {
+@Controller(['game-servers/static', 'game-servers'])
```
Comment (c43721) [inline src/game-servers/providers/static-game-server/controllers/static-game-servers.controller.ts]: 🤯 woah

File: src/game-servers/providers/static-game-server/models/static-game-server.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  @Exclude({ toPlainOnly: true })
+  @Prop()
+  internalIpAddress: string;
+
+  @Exclude({ toPlainOnly: true })
+  @Prop({ required: true })
+  rconPassword!: string;
+
+  @Prop({ default: false })
+  isOnline!: boolean; // was the server online last we checked
```
Comment (c43721) [inline src/game-servers/providers/static-game-server/models/static-game-server.ts]: Use proper doc comments (like the one in InteralIpAddress)

File: src/game-servers/providers/static-game-server/models/static-game-server.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Prop({ default: true })
+  isClean!: boolean; // is the server cleaned up after the last game
```
Comment (c43721) [inline src/game-servers/providers/static-game-server/models/static-game-server.ts]: proper doc comments

File: src/game-servers/providers/static-game-server/models/static-game-server.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  lastHeartbeatAt?: Date;
+
+  @Prop({ default: 1 })
+  priority!: number;
+
+  @Prop()
+  customVoiceChannelName?: string;
+
+  async rcon(): Promise<Rcon> {
+    return createRcon({
```
Comment (c43721) [inline src/game-servers/providers/static-game-server/models/static-game-server.ts]: can we await this?

**Review by c43721 — CHANGES_REQUESTED** [review]

File: src/game-servers/models/game-server.ts (hunk 1 above)
Comment (garrappachc) [reply src/game-servers/models/game-server.ts]: I wanted to make this abstract, but mongoose (or rather NestJS) won't let me create a schema out of an abstract class.

File: src/game-servers/providers/static-game-server/controllers/static-game-servers.controller.ts (hunk 2 above)
Comment (garrappachc) [reply src/game-servers/providers/static-game-server/controllers/static-game-servers.controller.ts]: Yeah, I'll make some kind of proxy controller that marks all of `game-servers` subroutes deprecated.

**Comment by garrappachc** [comment]
@c43721 review changes please

File: src/game-servers/providers/static-game-server/models/static-game-server.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  rconPassword!: string;
+
+  /**
+   * Was the server online last time we checked?
+   */
+  @Prop({ default: false })
+  isOnline!: boolean;
+
+  /**
+   * Ss the server cleaned up after the last game?
```
Comment (c43721) [inline src/game-servers/providers/static-game-server/models/static-game-server.ts]: Typo

**Review by c43721 — CHANGES_REQUESTED** [review]

File: src/game-servers/providers/static-game-server/models/static-game-server.ts (hunk 6 above)
Comment (garrappachc) [reply src/game-servers/providers/static-game-server/models/static-game-server.ts]: Fixed

File: src/game-servers/providers/static-game-server/models/static-game-server.ts [hunk 7]
```diff
@@ -28,7 +28,7 @@ export class StaticGameServer extends GameServer {
   isOnline!: boolean;
 
   /**
-   * Ss the server cleaned up after the last game?
+   * Is the server cleaned up after the last game?
```
Comment (c43721) [inline src/game-servers/providers/static-game-server/models/static-game-server.ts]: Proper grammar says to use "was" 🙂

**Review by c43721 — CHANGES_REQUESTED** [review]

File: src/game-servers/providers/static-game-server/models/static-game-server.ts (hunk 7 above)
Comment (c43721) [reply src/game-servers/providers/static-game-server/models/static-game-server.ts]: the nittiest of the pickiest, they call me

Comment (garrappachc) [reply src/game-servers/providers/static-game-server/models/static-game-server.ts]: Nice 😇 Fixed.

**Review by c43721 — APPROVED** [review]

**Pull request merged**